### DISCUSSION TITLE
Syntax for unit tests

### DISCUSSION THREAD

**Discussion by applejag** [deepnoodle-ai/risor#127] [Ideas]
Would be nice to have. I find myself wanting to write tests for Bash scripts, but too tedious to do. Writing tests in Python scripts is quite easy and painless.

I propose having tests be able to be defined in the same file, like how [Rust](https://doc.rust-lang.org/book/ch11-01-writing-tests.html) and [Zig](https://ziglang.org/documentation/master/#Zig-Test) allows you to do.

Something like this would be nice:

```go
// File: my-script.risor

func add(a, b) {
  return a + b
}

// defines test named "my-script.risor"
test {
  assert(add(1,2) == 3)
}

// defines test named "my-script.risor/add fails"
test "add fails" {
  assert(add(1,2) == 4)
}
```

Some design concerns:

## What about top-level statements

```go
print("Hello")

test {
  assert(1 == 1)
}
```

Will the `print` run on every test? Only run once? Not run at all?

For this I'd say could take inspiration from Python's `__name__ == "__main__"`, but simplified to maybe "if a `main()` function exists, then run that after loading the script file". So it'll be:

```go
print("Run once at 'risor my-script.risor' and 'risor test my-script.risor'")

func main() {
  print("Only runs when 'risor my-script.risor'")
}

test {
  print("Only runs when 'risor test my-script.risor'")
  assert(1 == 1)
}
```

Alternatively, could be some global variable from a module like:

```go
print("Run once at 'risor my-script.risor' and 'risor test my-script.risor'")

if script.is_running {
  print("Only runs when 'risor my-script.risor'")
}

if script.is_testing {
  print("Only runs when 'risor test my-script.risor'")
}

test {
  assert(1 == 1)
}
```

## What about test init/cleanup

For starters, just force the user be explicit. If this becomes a main pain-point, then revisit this.

```go
func test_init() {
  print("Do some init thing for these tests")
}

func test_cleanup() {
  print("Do some init thing for these tests")
}

test {
  test_init()
  assert(1 == 1)
  test_cleanup() // maybe good idea to introduce a 'defer' keyword?
}

test {
  test_init()
  assert(3 == 4)
  test_cleanup()
}
```

**Comment by myzie** [comment]
I'm supportive of the core idea. It seems like test declarations would basically be equivalent to function declarations. We could just add a `risor test` command to the CLI that loads module(s) and then runs all the tests defined in the module.
